flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/temporal/TemporalRowTimeJoinOperator.java

```diff
@@ -0,0 +1,376 @@
+import org.apache.flink.api.common.state.MapState;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.VoidNamespace;
+import org.apache.flink.runtime.state.VoidNamespaceSerializer;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.InternalTimer;
+import org.apache.flink.streaming.api.operators.InternalTimerService;
+import org.apache.flink.streaming.api.operators.TimestampedCollector;
+import org.apache.flink.streaming.api.operators.Triggerable;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.dataformat.BaseRow;
+import org.apache.flink.table.dataformat.JoinedRow;
+import org.apache.flink.table.dataformat.util.BaseRowUtil;
+import org.apache.flink.table.generated.GeneratedJoinCondition;
+import org.apache.flink.table.generated.JoinCondition;
+import org.apache.flink.table.typeutils.BaseRowTypeInfo;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This operator works by keeping on the state collection of probe and build records to process
+ * on next watermark. The idea is that between watermarks we are collecting those elements
+ * and once we are sure that there will be no updates we emit the correct result and clean up the
+ * state.
+ *
+ * <p>Cleaning up the state drops all of the "old" values from the probe side, where "old" is defined
+ * as older then the current watermark. Build side is also cleaned up in the similar fashion,
+ * however we always keep at least one record - the latest one - even if it's past the last
+ * watermark.
+ *
+ * <p>One more trick is how the emitting results and cleaning up is triggered. It is achieved
+ * by registering timers for the keys. We could register a timer for every probe and build
+ * side element's event time (when watermark exceeds this timer, that's when we are emitting and/or
+ * cleaning up the state). However this would cause huge number of registered timers. For example
+ * with following evenTimes of probe records accumulated: {1, 2, 5, 8, 9}, if we
+ * had received Watermark(10), it would trigger 5 separate timers for the same key. To avoid that
+ * we always keep only one single registered timer for any given key, registered for the minimal
+ * value. Upon triggering it, we process all records with event times older then or equal to
+ * currentWatermark.
+ */
+public class TemporalRowTimeJoinOperator
```

Comment:
`TemporalRowTimeJoinOperator` logical is a bit complicated, consider add a `TemporalRowTimeJoinOperatorTest`?